The original file is provided by mpd, and
can be found in the Buildroot package directory.

Instead of creating an overlay file, having to check this manually whenever
changes are made within Buildroot, patch the master file with additional code

Upstream: configuration change

Signed-off-by: Andreas Ziegler <br025@umbiko.net>
---
--- a/etc/init.d/S95mpd	2026-01-24 07:55:04.732178754 +0100
+++ b/etc/init.d/S95mpd	2026-01-24 11:19:30.911692102 +0100
@@ -2,20 +2,93 @@
 #
 # S95mpd	Starts Music Player daemon.
 #
+# shellcheck source=/dev/null # do not follow included files
 # shellcheck disable=SC2329 # functions are called via variable
 
 DAEMON="mpd"
-PIDFILE="/var/run/$DAEMON.pid"
+PIDFILE="/var/run/$DAEMON/pid"
 
 # Sanity checks
 [ -f /etc/$DAEMON.conf ] || exit 0
+[ -f /usr/bin/$DAEMON ] || exit 0
+
+HOTPLUG_DIR=/media
+
+MPD_LIB="/var/lib/$DAEMON"
+MPD_LOG="/var/log/$DAEMON"
+MPD_RUN=$(dirname $PIDFILE)
+
+MPD_MUSIC=$MPD_LIB/music
+MPD_CACHE=$MPD_LIB/cache
+
+IO_URING_MEMLIMIT=65536
+
+### initialize system variables
+[ -f /etc/nabla.conf ] && . /etc/nabla.conf
+
+### check if $DAEMON is configured to run
+[ "$NABLA_START_MUSICPLAYER" != "Y" ] && exit 0
+
+### functions that are used in more than one init script
+[ -x /etc/init.d/init_functions ] && . /etc/init.d/init_functions
+
+# unmute all alsa devices found on machine
+unmute() {
+	amixer="/usr/bin/amixer"
+	procfs="/proc/asound"
+
+	if [ -x "$amixer" ]; then
+		# loop through all sound devices
+		for card in "$procfs"/card[0-9]; do
+			# ignore non-USB devices
+			if [ ! -f "$card/usbid" ]; then continue; fi
+			# extract device id
+			id=${card##*/card}
+			# unmute all simple controls on this device
+			amixer -c "$id" scontrols | \
+				sed -e 's/^Simple mixer control//' | \
+				while read -r line; do 
+					amixer -c "$id" sset "$line" 100% unmute > /dev/null 2>&1
+				done
+		done
+	fi
+}
+
+# create directories and link to music libraries
+initialize() {
+	_owner="mpd:audio"
+	mkdir -p "$MPD_LIB" && chown -R "$_owner" "$MPD_LIB"
+	mkdir -p "$MPD_CACHE" && chown -R "$_owner" "$MPD_CACHE"
+	mkdir -p "$MPD_LOG" && chown -R "$_owner" "$MPD_LOG"
+	mkdir -p "$MPD_RUN" && chown -R "$_owner" "$MPD_RUN"
+	# this is the $DAEMON link to the music directory
+	if [ ! -h "$MPD_MUSIC" ]; then
+		ln -s "$HOTPLUG_DIR" "$MPD_MUSIC"
+	fi
+	# create ignore list
+	echo "$NABLA_PARTITION" > "$HOTPLUG_DIR"/.mpdignore
+	# library file attributes may be broken by import
+	find "$MPD_LIB" -type f -exec chmod u=rw,g=rw,o=r {} \;
+}
 
 start() {
 	printf "Starting %s: " "$DAEMON"
+	initialize
+	unmute
+	# increase max locked memory limit for io_uring 
+	# increase real-time priority limit to enable real-time scheduling
+	# shellcheck disable=SC3045 # busybox sh provides ulimit builtin
+	ulimit -HS -l "$IO_URING_MEMLIMIT" -r $((HARDIRQ_PRIORITY-1))
 	start-stop-daemon --start --pidfile "$PIDFILE" \
 		--exec "/usr/bin/$DAEMON"
 	status=$?
 	if [ "$status" -eq 0 ]; then
+		# set processor affinity
+		cpu_id=$(_get_rt_cpuid)
+		pid=$(cat "$PIDFILE")
+		taskset -p "$(_get_cpumask "$cpu_id")" "$pid"
+		# thread priority is set by process, so this might be unnecessary
+		chrt "$(_chrt_policy)" -p "$SOFTIRQ_PRIORITY" "$pid"
 		echo "OK"
 	else
 		echo "FAIL"
